File: app/src/main/java/fr/iutlens/mmi/rogue/RandomChooser.java
package fr.iutlens.mmi.rogue;

import android.os.Build;
import android.support.annotation.RequiresApi;

import java.util.Iterator;
import java.util.List;
import java.util.Random;
import java.util.Vector;
import java.util.stream.Collectors;

/**
 * Created by dubois on 24/12/2019.
 */

class RandomChooser<T> {
    private Random rand = new Random();

    private class Choice{

        private final T description;
        private final int size;
        private final int weight;

        public Choice(T description, int size, int weight) {
            this.description = description;
            this.size = size;
            this.weight = weight;
        }
    }

    Vector<Choice> v;

    RandomChooser(){
        v = new Vector<>();
    }

    RandomChooser(T[] list){
        v = new Vector<>();

        for(T t : list){
            addOption(t,1,1);
        }
    }

    public void addOption(T description, int size, int weight) {
        v.add(new Choice(description,size,weight));
    }

    private Choice getOne(Vector<Choice> possible){
        int sumW = 0;
        for(Choice c : v) sumW+= c.weight;
        if (sumW <=0) return null; //FIXME
        sumW = rand.nextInt(sumW);
        for(Choice c :possible){
            sumW -=c.weight;
            if (sumW <0){
                return c;
            }
        }
        return null;
    }

    public T getOne(){
        final Choice choice = getOne(v);
        if (choice==null) return null;
        return choice.description;
    }

    public T getOne(T except){
        Vector<Choice> possible = new Vector<>(v);

        Iterator it = possible.iterator();
        while (it.hasNext()){
            if (except.equals(it.next())) it.remove();
        }
        return getOne(possible).description;
    }


    public Vector<T> get(int maxSize){
        Vector<T> result = new Vector<>();
        Vector<Choice> possible = new Vector<>();

        for(Choice c : v) if (c.size <= maxSize) possible.add(c);

        while (maxSize >0 && possible.size()>0) {
            Choice c = getOne(possible);
            if (c == null) return result; // FIXME
            result.add(c.description);
            maxSize -= c.size;


            Iterator<Choice> it = possible.iterator();
            while(it.hasNext()) if (it.next().size> maxSize) it.remove();
        }
        return  result;
    }
}

File: app/src/main/java/fr/iutlens/mmi/rogue/RandomChooser.kt
package fr.iutlens.mmi.rogue

import android.util.Pair
import java.util.*

/**
 * Created by dubois on 24/12/2019.
 */
class RandomChooser<T> {
    private val rand = Random()

    private inner class Choice(val description: T, val size: Int, val weight: Int)

    private var v: Vector<Choice>

    constructor() {
        v = Vector()
    }

    constructor(list: Array<T>) {
        v = Vector()
        for (t in list) {
            addOption(t, 1, 1)
        }
    }

    fun addOption(description: T, size: Int, weight: Int) {
        v.add(Choice(description, size, weight))
    }

    private fun getOne(possible: Vector<Choice>): Choice? {
        var sumW = 0
        for (c in v) sumW += c.weight
        if (sumW <= 0) return null //FIXME
        sumW = rand.nextInt(sumW)
        for (c in possible) {
            sumW -= c.weight
            if (sumW < 0) {
                return c
            }
        }
        return null
    }

    val one: T?
        get() {
            val choice = getOne(v) ?: return null
            return choice.description
        }

    fun getOne(except: T): T {
        val possible = Vector(v)
        val it: MutableIterator<*> = possible.iterator()
        while (it.hasNext()) {
            if (except == it.next()) it.remove()
        }
        return getOne(possible)!!.description
    }

     fun get(maxSize: Int): Vector<T> {
        var maxSize = maxSize
        val result = Vector<T>()
        val possible = Vector<Choice>()
        for (c in v) if (c.size <= maxSize) possible.add(c)
        while (maxSize > 0 && possible.size > 0) {
            val c = getOne(possible) ?: return result
            // FIXME
            result.add(c.description)
            maxSize -= c.size
            val it = possible.iterator()
            while (it.hasNext()) if (it.next().size > maxSize) it.remove()
        }
        return result
    }
}